Overhaul

bedeoverend opened the pull request:
This changes the behavior to be a factory, rather than a general behavior. This means that the behavior is specific for the element, and `setupSimpla` is no longer needed e.g.

```js
Polymer({
  is: 'my-element',

  behaviors: [ SimplaBehaviors.Element ],

  created() {
    this.setupSimpla(config);
  }
}
```
becomes:
```js
Polymer({
  is: 'my-element',

  behaviors: [ SimplaBehaviors.Element(config) ]
}
```

it also adds in a few features / bug fixes:

- Doesn't call `setCallback` during a `getCallback` - this fixes infinite loop bugs with Simpla SDK
- Fixes calling `getCallback` on an old path. Just needed a tweak in the code
- Add 'static content' support. Sets own values to buffer if initial get of content is empty
- Only sets up path observer if element is attached
- Swallows invalid path errors and warns instead
- Allows no data properties to be passed in - instead just using events

Also adds in basic tests

bedeoverend commented:
@seaneking rather than a behavior factory function, a better option might be to just expose a behavior, and tell users how to override it - this is much more 'Polymer 1.0' way of extending behaviors:

```js
let customSimplaBehavior = {
  properties: {
    type: {
      type: String,
      value: 'Text'
    }
  },

  listeners: [
    'input': '_protectedSetCallback'
  ],

  observers: {
    'text': '_protectedSetCallback'
  },

  _updateFromSimpla() {
    // Custom loading fn
  },

  _updateSimplaBuffer() {
    // Custom setting fn
  }
}

Polymer({
  is: 'my-element',

  properties: {
    text: String
  },

  behaviors: [ SimplaBehaviors.Element, customSimplaBehavior ]
});
```

I'm not sure on the specifics (e.g. the listeners and observers need to call special functions), but that's the general idea. That pattern is used much more in Polymer land. Once we move to Polymer 2 / any modern Web Component framework, we'd move to class mixins anyway, which would be much more similar to the above syntax.

seaneking commented:
Yeah I'd be keen to re-evaluate the general interface of this (ie: of how configuration is done), but that's probably better done when we move it to a base class / class mixin when WC 1.0 drops. For now I'd say let's just ship the factory behavior as a breaking change, because we'll need another breaking change when it becomes wc v1 ready anyway.

seaneking commented:
And yeah, LGTM 👍 

Assume you've tested this locally with existing elements (img, video, link)?

madeleineostoja commented on src/simpla-element-behavior.html:
```diff
@@ -10,185 +10,218 @@
+          getCallback,
+          setCallback,
+          dataProperties,
+          updateOn
```
FWIW this is purely for `simpla-text` since it's `value` isn't a prop but a getter - I think it would be cleaner to enforce elements using this to set their values to legit props and just use observers. Is that viable for simpla-text?

bedeoverend replied:
It's viable, but personally I like the event hook - I'll probably be using it for repeaters too, as repeater is dealing with complex data. If it's just dealing with plain props then it's easy e.g. text / boolean / number. But if you're doing mutations on an objects props you have to either tie in Polymer's observing of subproperties, or use events. IMO events are good as it can give you fine grained control over when to actually trigger changes.

madeleineostoja replied:
In that case I'd still we should use deep observation of the properties. It's a much cleaner interface - "here's the property with my data, update Simpla whenever it changes" - it shouldn't matter if that change is in a simple or complex value, that's an implementation detail.

Re: simpla-text, regardless I think value should be a property rather than a getter, properties are a standard for wc, and it's more robust from an interface point of view

madeleineostoja replied:
And then there would be no need for events, because there's no reason to update Simpla if the data property didn't change. The only need for events is if the 'properties' you're giving this behavior aren't properties, as is the case with simpla-text. In that case you're misusing the behavior, IMO. It expects the things it's getting to be properties, not a getter funciton.

bedeoverend commented:
Sounds good. Not yet, but I'll test before I merge

bedeoverend commented:
@seaneking I've added a couple of more things:

- Only sets up paths observer if element is attached
- Swallows and warns about invalid path errors
- Allows no data properties to be passed, instead just relying on events

That last one was needed for simpla-text. Which means simpla-text is now also using custom getters / setters, same as simpla-img. It's cleaner than forcing simpla-text to use a property instead of a getters / setters. The getters / setters work really in well in that scenarios, and defining an observer on a data property seems to destroy the getters and setters. But, this does raise whether this behavior's interface is quite right. Not sure.

bedeoverend commented:
Perhaps something better would be being able to pass serialize / deserialize callbacks:

```js
SimplaBehaviors.Element({
  type: 'Link',
  serialize: '_serializeToSimpla',
  deserialize: '_deserializeFromSimpla'
});

Polymer({
  is: 'simpla-img',
  
  serialize() {
    let src = this.src || DEFAULT_SRC,
        alt = this.alt;
        
    return { src, alt };
  },

  deserialize(item) {
    Object.assign(this, item.data);
  }
});
```

seaneking commented:
> Allows no data properties to be passed, instead just relying on events

I _strongly_ disagree with this, and these commits should be rolled back. You're changing how this behavior works to suit simpla-text. Simpla-text should store it's value in a property, not in getters/setters. That is the standard for web components. This behavior expects properties, the fact that it destroys simpla-text's getters and setters is a good thing, it means that simpla-text is misusing it.

The whole point of this is 'here's the data properties of my element, sync them up to Simpla'. In any case, any drastic interface changes should be put off until we look at making this a class mixin for wc v1, but even then I would still strongly advocate for a property-based setup, since that's the gold standard for developing web components.

seaneking commented:
> Swallows and warns about invalid path errors

I thought we previously discussed to just let these throw, but clean up the error messages in the SDK? The warnings where there previously because the errors were terrible. I'd say leave this out of this behavior, since the errors will only bust the `get` promise, and not blow the developer's whole stack anyway.

seaneking commented:
> Perhaps something better would be being able to pass serialize / deserialize callbacks

This is kind of like what we have now with `_getCallback` and `_setCallback`, just with defaults already defined in the behavior. 

But yeah I'd be keen to abstract this into general functions like the examples in your snippet, which we use inside `get`/`set`, but still with defaults in the behavior so it continues to work out of the box for the common case. Then elements wouldn't have to overwrite the entire callback if they have to do special things, just the property-specific logic (in your snippet, just the set function). Also I think `set` and `get` a much better option/function names here, mirroring how they're used.

bedeoverend commented:
Re: properties / bindings. Happy to give it a shot to get simpla-text to work purely with properties, not sure how successful it will be though, and it would be a real shame to not be able to use this behavior given how much boilerplate it gives you. Worth noting that properties over getters / setters is a very Polymer way of building web components (given that this is a polymer behavior and that's what simpla-text is using, fair enough) but getters and setters certainly have their place in WebComponents in general.

 > I thought we previously discussed to just let these throw, but clean up the error messages in the SDK

Wasn't aware of that, thought we were still including these, but that's fine. I'll revert those commits.

>This is kind of like what we have now with _getCallback and _setCallback, just with defaults already defined in the behavior.

Yup, just trying to DRY them out a bit. If you're keen to do this now, happy with this syntax?

```js
SimplaBehavior.Element({
  type: 'Text',
  set: '_setToSimpla',
  get: '_getFromSimpla'
});
```

`_setToSimpla` expects a `data` object to be returned, and `_getFromSimpla` is given an `item`?

bedeoverend commented:
Have rolled back those last two commits

seaneking commented:
👍 re: that last point, happy with that - though as long as we still provide default functions in the behavior (same as current, just object assign back and forth). IMO it's good just to be able to more easily overwrite the bits that would need to be overwritten for more complex use-cases.

Re: props - I was going off that 'gold standard for web components' doc you showed me a while ago, which I agreed with. Though I guess that was written with heavy input from polymer people. Still, I personally can't see a benefit of getters/setters over properties for something like an element's value.

bedeoverend commented:
Yup w/ default functions that's good 👍  I'll implement

Re: props - I don't think they were against getters / setters? I mean, a property is still a property if it has getters and setters, it's just accessed differently. In the early polymer days the guy who wrote the gold-standards wiki was actually [requesting](https://github.com/Polymer/polymer/issues/2454) better support on getters / setters (though that may be me projecting on that issue 😉). Bidelman also [used](https://developers.google.com/web/fundamentals/getting-started/primers/customelements#attrchanges) them when introducing v1 Custom Elements. Having a getter / setter doesn't mean the property will be any different to the consumer, it's just that you can't use them with Polymer's observers / data-binding systems.

bedeoverend commented:
Have made those changes, removed the `updateOn` event, given that it's not being used anymore, and updated the README.

@seaneking could you give this a look over and LGTM? Feel free to change if needed.

bedeoverend commented:
@seaneking see that last commit, what do you think?

Basically we need a way for a dev to bail out of the set callback e.g. simpla-img doesn't want to set to simpla if the `src` is placeholder. To do this I've set it so that if the `set` callback returns undefined, it will stop. Another option (popular in react land) is passing an optional `shouldSet` callback which should return a boolean of whether it should set. The behavior would call that before any call to `set`.

seaneking commented:
I think that's fine - though I'd probably preference bailing on `undefined || null`. Re: `shouldSet`, I think it's too much of an edge case to have to pass a callback on every set no? If it was a 50/50 thing then sure, but this is a rare use-case.

Re: props and getters/setters, maybe I've misunderstood the innards of wc v1, but you couldn't easily serialize a getter/setter to a legit attribute on the element could you?

seaneking commented:
And when we rebuild this as an ES6 mixin/base class, I'd be very keen to do our own deep observing (using ES6 proxies?) rather than leaning on polymer. There's no reason this should need to be tied to a specific framework, all we'd be doing is "when the value of this property in my class changes, run these callbacks".

Not sure how that affects the setters/getters discussion.

seaneking commented:
Also now that I look at it actually being used, the `set` and `get` names for those hooks is a little unintuitive, what about just keeping the old names, `setCallback` and `getCallback`? At least it's obvious that they're functions then.

bedeoverend commented:
Happy to leave do `undefined || null`. `shouldSet` wouldn't pass a callback on every set, you'd just give a `shouldSet` function in the options e.g.

```js
SimplaBehaviors.Element({
  setCallback: '_setToSimpla',
  shouldSet: '_shouldSetToSimpla'
});
```

and if one isn't passed, it would always set. But honestly I'm happy to just leave this off, can add it later if we want.

Re: discussing stuff around getters / setters + WC class mixin. Think that's getting off topic for now, lets address / discuss that further down the track when we need to.

Re: `setCallback` and `getCallback` yup that's good, was going to suggest the same thing.

seaneking commented:
Coolio - so is this good to be merged? Are we going to deal with complex data (ie: deep observing for collections) or in the SDK?

bedeoverend commented:
Yeah this can be merged, I'll add in the complex data afterwards, but that can be a minor (patch rather), as it's backwards compatible. It'll need to be done here, but it'll also rely on a patch in the SDK for it to work properly.

There are a few other changes that need to be made, but may as well do them later / separate PR: shouldn't start two-way sync until after the first get is made; shouldn't two-way sync while detached / attached. They should be patches IMO as otherwise it's unintended / buggy behavior. If it's a break patch can just deal with semver anyway though.
